internal/controller/compaction/reconciler.go

```diff
@@ -359,6 +353,129 @@ func (r *Reconciler) createCompactionJob(ctx context.Context, logger logr.Logger
 	return job, nil
 }
 
+func handleFailedJob(ctx context.Context, r *Reconciler, logger logr.Logger, job *batchv1.Job, etcd *druidv1alpha1.Etcd, jobCompletionReason string) error {
+	switch jobCompletionReason {
+	case batchv1.JobReasonDeadlineExceeded:
+		logger.Info("Job has been completed due to deadline exceeded", "namespace", job.Namespace, "name", job.Name)
+		metricJobsTotal.With(prometheus.Labels{druidmetrics.LabelSucceeded: druidmetrics.ValueSucceededDeadlineExceeded, druidmetrics.EtcdNamespace: etcd.Namespace}).Inc()
+		metricJobDurationSeconds.With(prometheus.Labels{druidmetrics.LabelSucceeded: druidmetrics.ValueSucceededDeadlineExceeded, druidmetrics.EtcdNamespace: etcd.Namespace}).Observe(float64(*job.Spec.ActiveDeadlineSeconds))
+	case batchv1.JobReasonBackoffLimitExceeded:
+		if err := handleBackoffLimitExceeded(ctx, r, logger, job, etcd); err != nil {
+			return fmt.Errorf("error while handling job's failure condition type backoffLimitExceeded: %v", err)
+		}
+	default:
+		logger.Info("Job has been completed with unknown reason", "namespace", job.Namespace, "name", job.Name)
+		metricJobsTotal.With(prometheus.Labels{druidmetrics.LabelSucceeded: druidmetrics.ValueSucceededUnknown, druidmetrics.EtcdNamespace: etcd.Namespace}).Inc()
+		metricJobDurationSeconds.With(prometheus.Labels{druidmetrics.LabelSucceeded: druidmetrics.ValueSucceededUnknown, druidmetrics.EtcdNamespace: etcd.Namespace}).Observe(time.Since(job.Status.StartTime.Time).Seconds())
+	}
+
+	if err := r.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+		return fmt.Errorf("error while deleting failed compaction job: %v", err)
+	}
+	return nil
+}
+
+func handleBackoffLimitExceeded(ctx context.Context, r *Reconciler, logger logr.Logger, job *batchv1.Job, etcd *druidv1alpha1.Etcd) error {
+	logger.Info("Job has been completed due to backoffLimitExceeded", "namespace", job.Namespace, "name", job.Name)
+	pod, err := getPodForJob(ctx, r.Client, logger, job)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Couldn't fetch pods for job", "namespace", job.Namespace, "name", job.Name)
+			return fmt.Errorf("error while fetching pod for job: %v", err)
+		}
+		logger.Info("No pods found for job", "namespace", job.Namespace, "name", job.Name)
+	}
+	podFailureReason, lastTransitionTime := getPodFailureReasonAndLastTransitionTime(pod)
+	handlePodFailureReason(logger, pod, etcd, podFailureReason, lastTransitionTime, job.Status.StartTime.Time)
```

Comment:
All this function does is record a metric. Why can it not have a more appropriate name?